packages/vats/src/lib-board.js

```diff
@@ -28,28 +61,114 @@ const calcCrc = (data, crcDigits) => {
 };
 
 /**
- * Create a board to post things on.
+ * Make board state init function, closing over baggage.
  *
- * @param {bigint | number} [initSequence]
- * @param {object} [options]
- * @param {string} [options.prefix]
- * @param {number} [options.crcDigits]
+ * @see {prepareExoClassKit}
+ * @see {@link ../../SwingSet/docs/virtual-objects.md|SwingSet Virtual Objects}
+ *
+ * Hoisting this function makes defining the state type concise.
+ * @typedef {ReturnType<ReturnType<makeInit>>} BoardState
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ *
+ * // TODO: use Key from @agoric/store when available
+ * @typedef {import('@endo/marshal').Passable} Key
  */
-function makeBoard(
-  initSequence = 0,
-  { prefix = DEFAULT_PREFIX, crcDigits = DEFAULT_CRC_DIGITS } = {},
-) {
+const makeInit =
+  baggage =>
+  /**
+   * @param {bigint | number} [initSequence]
+   * @param {object} [options]
+   * @param {string} [options.prefix] prefix for all ids generated
+   * @param {number} [options.crcDigits] count of digits to use in CRC at end of the id
+   */
+  (
+    initSequence = 0,
+    { prefix = DEFAULT_PREFIX, crcDigits = DEFAULT_CRC_DIGITS } = {},
+  ) => {
+    const immutable = { prefix, crcDigits };
+
+    const lastSequence = BigInt(initSequence);
+    /** @type {MapStore<string, Key>} */
+    const idToVal = provideDurableMapStore(baggage, 'idToVal');
+    /** @type {MapStore<Key, string>} */
+    const valToId = provideDurableMapStore(baggage, 'valToId');
+    const mutable = { lastSequence, idToVal, valToId };
+
+    return {
+      ...immutable,
+      ...mutable,
+    };
+  };
+
+// #region Marshaller utils
+// These utils are used by the {readonlyMarshaller, publishingMarshaller} facets
+// to serialize/unserialize. They can't return the marshallers themselves
+// because of the heap cost of maintaining those objects. Instead they make
+// transient marshallers that get GCed when the function completes.
+
+/**
+ * @param {Key} value
+ * @param {BoardState} state
+ */
+const getId = (value, state) => {
+  const { idToVal, valToId, prefix, crcDigits } = state;
+  if (!valToId.has(value)) {
+    state.lastSequence += 1n;
+    const seq = state.lastSequence;
+
+    const crcInput = `${prefix}${seq}`;
+    const crc = calcCrc(crcInput, crcDigits);
+
+    const id = `${prefix}${crc}${seq}`;
+
+    valToId.init(value, id);
+    idToVal.init(id, value);
+    return id;
+  }
+  return valToId.get(value);
+};
+
+/**
+ * @param {string} id
+ * @param {Readonly<BoardState>} state
+ */
+const getValue = (id, { prefix, crcDigits, idToVal }) => {
+  id.startsWith(prefix) || Fail`id must start with ${prefix}: ${id}`;
```

Comment:
Should we really be censoring `prefix`?
```suggestion
  id.startsWith(prefix) || Fail`id must start with ${q(prefix)}: ${id}`;
```